### DISCUSSION TITLE
Install triggers auto-cleanup triggers upgrade?

### DISCUSSION THREAD

**Discussion by jvf** [Homebrew/discussions#101] [Everyday usage] [answered]
I tried to install a formula with `brew install <formula>` (`brew install ipython` in my case). It went like this (shortened output):

```
$ brew install ipython
Updating Homebrew...
==> Auto-updated Homebrew!
[...]
==> Installing dependencies for ipython: openssl@1.1, python@3.9 and zeromq
[...]
==> Installing ipython
[...]
==> Installing ipython
==> Pouring ipython-7.18.1_1.mojave.bottle.tar.gz
Error: The `brew link` step did not complete successfully
[...]
==> `brew cleanup` has not been run in 30 days, running now...
[...]
==> Upgrading 18 dependents:
[...] # proceeds to upgrade many major software packages, takes a long time
```

Auto updating brew and auto cleanup I understand, but how or why a complete upgrade is triggered in this process I do not understand. Is this intended behaviour? How can I prevent this?

**Comment by jonchang** [comment, marked as answer]
This is expected, though perhaps surprising, behavior.

* `brew install ipython`
* `ipython` needs `openssl@1.1`
* `openssl@1.1` is outdated, so it gets upgraded
* All reverse dependents of `openssl@1.1` get upgraded

Homebrew does not support arbitrary mixing and matching of package versions, so everything within a given package "tree" will need to be updated to the latest version as we do not test all possible combinations of versions. This is quite common behavior for package managers.

The messaging could better explain this, either in our documentation or at `brew upgrade` time, but this is not a bug in Homebrew. The `cleanup` message is also a red herring (this could run at any time, really).

**Reply by jvf** [reply]
Thank you for the explanation, very helpful!

Would it have helped to `brew pin` a reverse dependent of `openssl@1.1` here? Assuming the pinned version is compatible with the updated `openssl@1.1`.

**Reply by SMillerDev** [reply]
`brew pin` could have stopped this update, but it's mostly just a great way to introduce very hard to debug issues.

**Reply by jvf** [reply]
What is the recommended approach then? There does not seem to be a way to dry-run `brew install`, how can I know that an install will trigger the upgrade of another (not directly dependent) formula?

**Reply by SMillerDev** [reply]
You can't, the recommended approach is just to run `brew upgrade` frequently so updates happen on their own moment. And otherwise you'll just need to accept that it will update dependencies when it installs something.

**Comment by SMillerDev** [comment]
This is intended behaviour. Homebrew software is only build and tested with the latest dependencies. if `ipython` depends on `A 2.0` but you have `A 1.1` it isn't tested and will very likely be broken. That's why Homebrew updates all of the dependencies when you install something.
